hack/lib/golang.sh

```diff
@@ -87,77 +131,67 @@ readonly KUBE_NODE_TARGETS
 readonly KUBE_NODE_BINARIES=("${KUBE_NODE_TARGETS[@]##*/}")
 readonly KUBE_NODE_BINARIES_WIN=("${KUBE_NODE_BINARIES[@]/%/.exe}")
 
-if [[ -n "${KUBE_BUILD_PLATFORMS:-}" ]]; then
-  IFS=" " read -ra KUBE_SERVER_PLATFORMS <<< "${KUBE_BUILD_PLATFORMS}"
-  IFS=" " read -ra KUBE_NODE_PLATFORMS <<< "${KUBE_BUILD_PLATFORMS}"
-  IFS=" " read -ra KUBE_TEST_PLATFORMS <<< "${KUBE_BUILD_PLATFORMS}"
-  IFS=" " read -ra KUBE_CLIENT_PLATFORMS <<< "${KUBE_BUILD_PLATFORMS}"
-  readonly KUBE_SERVER_PLATFORMS
-  readonly KUBE_NODE_PLATFORMS
-  readonly KUBE_TEST_PLATFORMS
-  readonly KUBE_CLIENT_PLATFORMS
-elif [[ "${KUBE_FASTBUILD:-}" == "true" ]]; then
-  readonly KUBE_SERVER_PLATFORMS=(linux/amd64)
-  readonly KUBE_NODE_PLATFORMS=(linux/amd64)
-  if [[ "${KUBE_BUILDER_OS:-}" == "darwin"* ]]; then
-    readonly KUBE_TEST_PLATFORMS=(
-      darwin/amd64
-      linux/amd64
-    )
-    readonly KUBE_CLIENT_PLATFORMS=(
-      darwin/amd64
-      linux/amd64
-    )
+# Prints a space-separated string containing one of each duplicated element
+# in items
+kube::golang::duplicates() {
+  local items=$@
+  echo $(printf "%s\n" ${items[@]} | sort | uniq -d)
+}
+
+kube::golang::setup_platforms() {
+  if [[ -n "${KUBE_BUILD_PLATFORMS:-}" ]]; then
+    # Ensure each build platform is only specified once via sort -u.
+    # We use printf to insert newlines, which are required by sort.
+    # We then use echo to translate back to space-separated strings.
+    # This is to make sure the trick we do with kube::golang::duplicates is
+    # not foiled by duplicates in user input.
+    KUBE_BUILD_PLATFORMS=$(echo $(printf "%s\n" ${KUBE_BUILD_PLATFORMS[@]} | sort -u))
+
+    # Use kube::golang::duplicates to restrict the builds to the intersection
+    # with SUPPORTED_KUBE_*_PLATFORMS.
+    local server=$(kube::golang::duplicates "${KUBE_BUILD_PLATFORMS[@]} ${SUPPORTED_KUBE_SERVER_PLATFORMS[@]}")
+    IFS=" " read -ra KUBE_SERVER_PLATFORMS <<< "${server}"
```

Comment:
This pattern is wonky.  If you want to be most proper, the `mapfile` I showed above might be best:

```
mapfile -t KUBE_SERVER_PLATFORMS <<< \
    $(kube::golang::duplicates \
        "${KUBE_BUILD_PLATFORMS[@]}" \
        "${SUPPORTED_KUBE_SERVER_PLATFORMS[@]}" \
        )
readonly KUBE_SERVER_PLATFORMS

Reply:
Done